# **1. Perkenalan Dataset**


Tahap pertama, Anda harus mencari dan menggunakan dataset dengan ketentuan sebagai berikut:

1. **Sumber Dataset**:  
   Dataset dapat diperoleh dari berbagai sumber, seperti public repositories (*Kaggle*, *UCI ML Repository*, *Open Data*) atau data primer yang Anda kumpulkan sendiri.


# **2. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning atau deep learning.

In [ ]:
#Type your code here
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder

# **3. Memuat Dataset**

Pada tahap ini, Anda perlu memuat dataset ke dalam notebook. Jika dataset dalam format CSV, Anda bisa menggunakan pustaka pandas untuk membacanya. Pastikan untuk mengecek beberapa baris awal dataset untuk memahami strukturnya dan memastikan data telah dimuat dengan benar.

Jika dataset berada di Google Drive, pastikan Anda menghubungkan Google Drive ke Colab terlebih dahulu. Setelah dataset berhasil dimuat, langkah berikutnya adalah memeriksa kesesuaian data dan siap untuk dianalisis lebih lanjut.

Jika dataset berupa unstructured data, silakan sesuaikan dengan format seperti kelas Machine Learning Pengembangan atau Machine Learning Terapan

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
#Type your code here
student_performance_raw_df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Student_performance_data_raw.csv')
student_performance_raw_df.head()


,StudentID,Age,Gender,Ethnicity,ParentalEducation,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Extracurricular,Sports,Music,Volunteering,GPA,GradeClass
0,1001,17,1,0,2,19.833723,7,1,2,0,0,1,0,2.929196,2.0
1,1002,18,0,0,1,15.408756,0,0,1,0,0,0,0,3.042915,1.0
2,1003,15,0,2,3,4.210570,26,0,2,0,0,0,0,0.112602,4.0
3,1004,17,1,0,3,10.028829,14,0,3,1,0,0,0,2.054218,3.0
4,1005,17,1,0,2,4.672495,17,1,3,0,0,0,0,1.288061,4.0


# **4. Exploratory Data Analysis (EDA)**

Pada tahap ini, Anda akan melakukan **Exploratory Data Analysis (EDA)** untuk memahami karakteristik dataset.

Tujuan dari EDA adalah untuk memperoleh wawasan awal yang mendalam mengenai data dan menentukan langkah selanjutnya dalam analisis atau pemodelan.

In [ ]:
#Type your code here
# Melihat statistik data
student_performance_raw_df.describe(include='all')

,StudentID,Age,Gender,Ethnicity,ParentalEducation,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Extracurricular,Sports,Music,Volunteering,GPA,GradeClass
count,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000
mean,2196.500000,16.468645,0.510870,0.877508,1.746237,9.771992,14.541388,0.301421,2.122074,0.383361,0.303512,0.196906,0.157191,1.906186,2.983696
std,690.655244,1.123798,0.499986,1.028476,1.000411,5.652774,8.467417,0.458971,1.122813,0.486307,0.459870,0.397744,0.364057,0.915156,1.233908
min,1001.000000,15.000000,0.000000,0.000000,0.000000,0.001057,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1598.750000,15.000000,0.000000,0.000000,1.000000,5.043079,7.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,1.174803,2.000000
50%,2196.500000,16.000000,1.000000,0.000000,2.000000,9.705363,15.000000,0.000000,2.000000,0.000000,0.000000,0.000000,0.000000,1.893393,4.000000
75%,2794.250000,17.000000,1.000000,2.000000,2.000000,14.408410,22.000000,1.000000,3.000000,1.000000,1.000000,0.000000,0.000000,2.622216,4.000000
max,3392.000000,18.000000,1.000000,3.000000,4.000000,19.978094,29.000000,1.000000,4.000000,1.000000,1.000000,1.000000,1.000000,4.000000,4.000000


In [ ]:
# Melihat tipe data
student_performance_raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2392 entries, 0 to 2391
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   StudentID          2392 non-null   int64  
 1   Age                2392 non-null   int64  
 2   Gender             2392 non-null   int64  
 3   Ethnicity          2392 non-null   int64  
 4   ParentalEducation  2392 non-null   int64  
 5   StudyTimeWeekly    2392 non-null   float64
 6   Absences           2392 non-null   int64  
 7   Tutoring           2392 non-null   int64  
 8   ParentalSupport    2392 non-null   int64  
 9   Extracurricular    2392 non-null   int64  
 10  Sports             2392 non-null   int64  
 11  Music              2392 non-null   int64  
 12  Volunteering       2392 non-null   int64  
 13  GPA                2392 non-null   float64
 14  GradeClass         2392 non-null   float64
dtypes: float64(3), int64(12)
memory usage: 280.4 KB


In [ ]:
# Melihat missing values
student_performance_raw_df.isna().sum()

,0
StudentID,0
Age,0
Gender,0
Ethnicity,0
ParentalEducation,0
StudyTimeWeekly,0
Absences,0
Tutoring,0
ParentalSupport,0
Extracurricular,0


In [ ]:
# Melihat data duplikat
student_performance_raw_df.duplicated().sum()

np.int64(0)

***Insight***

- Total siswa adalah 2392. Student id seharusnya tidak dibutuhkan
- Umur rata-rata adalah 16 tahun dengan minimal 15 tahun dan maksimal 18 tahun.
- Gender masih belum direperesentasikan dengan baik karena bukan kategori. Akan tetapi sudah bagus untuk dimasukkan ke machine learning.
- Ethenicity (wilayah) belum direperesentasikan dengan baik. Akan tetapi sudah bagus untuk dimasukkan ke machine learning.Berdasarkan pemilik data, 0:Caucasian; 1: African American; 2: Asian 3: Other.
- ParentalEducation adalah level dari edukasi orang tua. Berdasarkan pemilik data, 0: None; 1: High School; 2: Some College; 3: Bachelor's; 4: Higher
- StudyTimeWeekly adalah waktu belajar perminggu. Rata-rata 9,7 jam dengan minimal hampir 0 jam dan maksimal 19,9 jam.
- Absences adalah total absen dengan rentan 0 - 30 jika berdasarkan pemberi data, 0-29 jika berdasarkan data.
- Tutoring: Klasifikasi biner, 0 jika tidak, 1 jika mengikuti tutor.
- Parental support: data ordinal terkait tingkat dukung orang tua dari 0 hingga 4.
-Extracurricular, Sports, Music, dan Volunteering: Klasifikasi biner, 1 jika mengikuti, 0 jika tidak mengikuti.
- GPA dan Gradeclass: GPA adalah  nilai yang dimiliki siswa dan GradeClass adalah kelompok nilainya (ini adalah variabel target). Pilih variabel target adalah GPA jika ingin hasil numerik yang lebih presisi. Pilih Gradeclass jika ingin hasil kategori secara langsung.

Pada kasus ini, variabel X yang akan digunakan adalah semua fitur kecuali GPA dan StudentID, Gender. Variabel target (Y) yang akan dicari adalah Gradeclass agar langsung ke inti dan mesin bisa belajar dengan lebih presisi.

Data yang digunakan adalah data yang sudah bersih, tetapi untuk mencegah di kemudian hari kurang bersihnya suatu data, maka akan tetap di preprocessing.

# **5. Data Preprocessing**

Pada tahap ini, data preprocessing adalah langkah penting untuk memastikan kualitas data sebelum digunakan dalam model machine learning.

Jika Anda menggunakan data teks, data mentah sering kali mengandung nilai kosong, duplikasi, atau rentang nilai yang tidak konsisten, yang dapat memengaruhi kinerja model. Oleh karena itu, proses ini bertujuan untuk membersihkan dan mempersiapkan data agar analisis berjalan optimal.

Berikut adalah tahapan-tahapan yang bisa dilakukan, tetapi **tidak terbatas** pada:
1. Menghapus atau Menangani Data Kosong (Missing Values)
2. Menghapus Data Duplikat
3. Normalisasi atau Standarisasi Fitur
4. Deteksi dan Penanganan Outlier
5. Encoding Data Kategorikal
6. Binning (Pengelompokan Data)

Cukup sesuaikan dengan karakteristik data yang kamu gunakan yah. Khususnya ketika kami menggunakan data tidak terstruktur.

In [ ]:
# Menangani data kosong
def tangani_missing_values(df):
  # Drop data kosong jika tidak memiliki ID
  df = df.dropna(subset=['StudentID'])

  '''
   Menangani kolom Age, ParentalSupport dan GPA dengan nilai rata-rata
   Menangani kolom Ethnicity kosong menjadi others
   Menangani kolom ParantalEducation kosong menjadi frekuensi tertinggi
   Menangani kolom StudyTimeWeekly dengan nilai 0 dengan asumsi bahwa tidak pernah belajar sehingga tidak mengisinya
   Menangani kolom Absences dengan nilai 0 dengan asumsi bahwa tidak pernah hadir sehingga tidak diabsen
   Menangani kolom Tutoring dengan klasifikasi 0 yaitu tidak mengikuti tutoring
    '''

  df = df.fillna({
    'Age': df['Age'].mean(),
    'ParentalSupport': df['ParentalSupport'].mean(),
    'GPA': df['GPA'].mean(),
    'Ethnicity': 3,
    'ParentalEducation': df['ParentalEducation'].mode()[0],
    'StudyTimeWeekly': 0,
    'Absences': 0,
    'Tutoring': 0
})
  return df

In [ ]:
# Menghapus data duplikat
def drop_duplicate(df):
  df.drop_duplicates(inplace=True)
  return df

In [ ]:
# Membuang kolom tidak diperlukan seperti StudentID dan Gender untuk mengatasi bias Gender
def drop_column(df):
  df.drop(['StudentID', 'Gender'], axis=1, inplace=True)
  return df

In [ ]:
# Deteksi dan penanganan outlier
def handle_outliers(df):
  # Mengubah Age menjadi rata-rata jika terdeteksi outlier
  df['Age'] = np.where((df['Age'] > df['Age'].mean() + 3 * df['Age'].std())| (df['Age'] <df['Age'].mean() - 3 * df['Age'].std()), df['Age'].mean(), df['Age'])
  # Mengubah ParentalSupport, ParantalEducation, ethnicity, StudyTimeWeekly, Absences, dan Tutoring sesuai rentan
  df['ParentalSupport'] = np.where(df['ParentalSupport'] > 4, 4, df['ParentalSupport'])
  df['ParentalSupport'] = np.where(df['ParentalSupport'] < 0, 0, df['ParentalSupport'])

  df['ParentalEducation'] = np.where(df['ParentalEducation'] > 4, 4, df['ParentalEducation'])
  df['ParentalEducation'] = np.where(df['ParentalEducation'] < 0, 0, df['ParentalEducation'])

  df['Ethnicity'] = np.where(df['Ethnicity'] > 3, 3, df['Ethnicity'])
  df['Ethnicity'] = np.where(df['Ethnicity'] < 0, 0, df['Ethnicity'])

  df['StudyTimeWeekly'] = np.where(df['StudyTimeWeekly'] > 20, 20, df['StudyTimeWeekly'])
  df['StudyTimeWeekly'] = np.where(df['StudyTimeWeekly'] < 0, 0, df['StudyTimeWeekly'])

  df['Absences'] = np.where(df['Absences'] > 30, 30, df['Absences'])
  df['Absences'] = np.where(df['Absences'] < 0, 0, df['Absences'])

  df['Tutoring'] = np.where(df['Tutoring'] > 1, 1, df['Tutoring'])
  df['Tutoring'] = np.where(df['Tutoring'] < 0, 0, df['Tutoring'])
  return df


In [ ]:
# Pembersihan data
student_performance_raw_df = tangani_missing_values(student_performance_raw_df)
student_performance_raw_df.describe()

,StudentID,Age,Gender,Ethnicity,ParentalEducation,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Extracurricular,Sports,Music,Volunteering,GPA,GradeClass
count,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000
mean,2196.500000,16.468645,0.510870,0.877508,1.746237,9.771992,14.541388,0.301421,2.122074,0.383361,0.303512,0.196906,0.157191,1.906186,2.983696
std,690.655244,1.123798,0.499986,1.028476,1.000411,5.652774,8.467417,0.458971,1.122813,0.486307,0.459870,0.397744,0.364057,0.915156,1.233908
min,1001.000000,15.000000,0.000000,0.000000,0.000000,0.001057,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1598.750000,15.000000,0.000000,0.000000,1.000000,5.043079,7.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,1.174803,2.000000
50%,2196.500000,16.000000,1.000000,0.000000,2.000000,9.705363,15.000000,0.000000,2.000000,0.000000,0.000000,0.000000,0.000000,1.893393,4.000000
75%,2794.250000,17.000000,1.000000,2.000000,2.000000,14.408410,22.000000,1.000000,3.000000,1.000000,1.000000,0.000000,0.000000,2.622216,4.000000
max,3392.000000,18.000000,1.000000,3.000000,4.000000,19.978094,29.000000,1.000000,4.000000,1.000000,1.000000,1.000000,1.000000,4.000000,4.000000


In [ ]:
student_performance_raw_df = drop_duplicate(student_performance_raw_df)
student_performance_raw_df.duplicated().sum()

np.int64(0)

In [ ]:
student_performance_raw_df = drop_column(student_performance_raw_df)
student_performance_raw_df.head()

,Age,Ethnicity,ParentalEducation,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Extracurricular,Sports,Music,Volunteering,GPA,GradeClass
0,17,0,2,19.833723,7,1,2,0,0,1,0,2.929196,2.0
1,18,0,1,15.408756,0,0,1,0,0,0,0,3.042915,1.0
2,15,2,3,4.210570,26,0,2,0,0,0,0,0.112602,4.0
3,17,0,3,10.028829,14,0,3,1,0,0,0,2.054218,3.0
4,17,0,2,4.672495,17,1,3,0,0,0,0,1.288061,4.0


In [ ]:
student_performance_raw_df = handle_outliers(student_performance_raw_df)
student_performance_raw_df.describe()

,Age,Ethnicity,ParentalEducation,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Extracurricular,Sports,Music,Volunteering,GPA,GradeClass
count,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000
mean,16.468645,0.877508,1.746237,9.771992,14.541388,0.301421,2.122074,0.383361,0.303512,0.196906,0.157191,1.906186,2.983696
std,1.123798,1.028476,1.000411,5.652774,8.467417,0.458971,1.122813,0.486307,0.459870,0.397744,0.364057,0.915156,1.233908
min,15.000000,0.000000,0.000000,0.001057,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,15.000000,0.000000,1.000000,5.043079,7.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,1.174803,2.000000
50%,16.000000,0.000000,2.000000,9.705363,15.000000,0.000000,2.000000,0.000000,0.000000,0.000000,0.000000,1.893393,4.000000
75%,17.000000,2.000000,2.000000,14.408410,22.000000,1.000000,3.000000,1.000000,1.000000,0.000000,0.000000,2.622216,4.000000
max,18.000000,3.000000,4.000000,19.978094,29.000000,1.000000,4.000000,1.000000,1.000000,1.000000,1.000000,4.000000,4.000000


MinmaxScaler = Age, Parental edu, Study time, absences, Parental support
One hot encode: Ethnicity

In [ ]:
# Menormalisasi Age, ParentalEducation, StudyTimeWeekly, Absences, dan PaerentalSupport
def normalize_features(df):
  scaler = MinMaxScaler()
  fitur = ['Age', 'ParentalEducation', 'StudyTimeWeekly', 'Absences', 'ParentalSupport']
  df[fitur] = scaler.fit_transform(df[fitur])
  return df

In [39]:
# Membuat one hot encode untuk ethnicity
def encode_feature(df):
  oh_encoder = OneHotEncoder(sparse_output=False, dtype=int)
  encoded_ethnicity = oh_encoder.fit_transform(df[['Ethnicity']])
  nama_etnis = ['Ethnicity_Caucasian', 'Ethnicity_African_American', 'Ethnicity_Asian', 'Ethnicity_Other']
  df_ethnicity_encoded = pd.DataFrame(encoded_ethnicity, columns=nama_etnis)
  df = pd.concat([df, df_ethnicity_encoded], axis=1)
  df.drop('Ethnicity', axis=1, inplace=True)
  return df

In [37]:
student_performance_raw_df = normalize_features(student_performance_raw_df)
student_performance_raw_df.head()

,Age,Ethnicity,ParentalEducation,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Extracurricular,Sports,Music,Volunteering,GPA,GradeClass
0,0.666667,0,0.50,0.992773,0.241379,1,0.50,0,0,1,0,2.929196,2.0
1,1.000000,0,0.25,0.771270,0.000000,0,0.25,0,0,0,0,3.042915,1.0
2,0.000000,2,0.75,0.210718,0.896552,0,0.50,0,0,0,0,0.112602,4.0
3,0.666667,0,0.75,0.501965,0.482759,0,0.75,1,0,0,0,2.054218,3.0
4,0.666667,0,0.50,0.233840,0.586207,1,0.75,0,0,0,0,1.288061,4.0


In [40]:
student_performance_raw_df = encode_feature(student_performance_raw_df)
student_performance_raw_df.head()

,Age,ParentalEducation,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Extracurricular,Sports,Music,Volunteering,GPA,GradeClass,Ethnicity_Caucasian,Ethnicity_African_American,Ethnicity_Asian,Ethnicity_Other
0,0.666667,0.50,0.992773,0.241379,1,0.50,0,0,1,0,2.929196,2.0,1,0,0,0
1,1.000000,0.25,0.771270,0.000000,0,0.25,0,0,0,0,3.042915,1.0,1,0,0,0
2,0.000000,0.75,0.210718,0.896552,0,0.50,0,0,0,0,0.112602,4.0,0,0,1,0
3,0.666667,0.75,0.501965,0.482759,0,0.75,1,0,0,0,2.054218,3.0,1,0,0,0
4,0.666667,0.50,0.233840,0.586207,1,0.75,0,0,0,0,1.288061,4.0,1,0,0,0
